dev-tools/omdb/src/bin/omdb/db.rs

```diff
@@ -1415,6 +1515,173 @@ async fn lookup_project(
+    opctx: &OpContext,
+    datastore: &DataStore,
+) -> Result<(), anyhow::Error> {
+    // A Crucible dataset is overprovisioned if size_used (amount taken up by
+    // Crucible region reservations) plus the control plane storage buffer
+    // (note this is _not_ a ZFS reservation! it's currently just a per-pool
+    // value in the database) is larger than the backing pool's total size.
+
+    let crucible_datasets =
+        datastore.crucible_dataset_list_all_batched(opctx).await?;
+
+    let Some(latest_collection) =
+        datastore.inventory_get_latest_collection(opctx).await?
+    else {
+        bail!("no latest inventory found!");
+    };
+
+    let mut zpool_total_size: HashMap<Uuid, i64> = HashMap::new();
+
+    for (_, sled_agent) in latest_collection.sled_agents {
+        for zpool in sled_agent.zpools {
+            zpool_total_size
+                .insert(zpool.id.into_untyped_uuid(), zpool.total_size.into());
+        }
+    }
+
+    let zpools: HashMap<Uuid, Zpool> = datastore
+        .zpool_list_all_external_batched(opctx)
+        .await?
+        .into_iter()
+        .map(|(zpool, _)| (zpool.id().into_untyped_uuid(), zpool))
+        .collect();
+
+    #[derive(Tabled)]
+    #[tabled(rename_all = "SCREAMING_SNAKE_CASE")]
+    struct CrucibleDatasetOverprovisionedRow {
+        // dataset fields
+        id: Uuid,
+        size_used: i64,
+        no_provision: bool,
+
+        // zpool fields
+        pool_id: Uuid,
+        control_plane_storage_buffer: i64,
+        pool_total_size: i64,
+    }
+
+    let mut rows = vec![];
+
+    for crucible_dataset in crucible_datasets {
+        if crucible_dataset.time_deleted().is_some() {
+            continue;
+        }
+
+        let control_plane_storage_buffer = zpools
+            .get(&crucible_dataset.pool_id)
+            .unwrap()
+            .control_plane_storage_buffer();
+
+        let total_dataset_size =
+            crucible_dataset.size_used + control_plane_storage_buffer;
+
+        match zpool_total_size.get(&crucible_dataset.pool_id) {
+            Some(pool_total_size) => {
+                if total_dataset_size >= *pool_total_size {
+                    rows.push(CrucibleDatasetOverprovisionedRow {
+                        id: crucible_dataset.id().into_untyped_uuid(),
+                        size_used: crucible_dataset.size_used,
+                        no_provision: crucible_dataset.no_provision(),
+
+                        pool_id: crucible_dataset.pool_id,
+                        control_plane_storage_buffer,
+                        pool_total_size: *pool_total_size,
+                    });
+                }
+            }
+
+            None => {}
+        }
+    }
+
+    let table = tabled::Table::new(rows)
+        .with(tabled::settings::Style::psql())
+        .with(tabled::settings::Padding::new(0, 1, 0, 0))
+        .to_string();
+
+    println!("{}", table);
+
+    Ok(())
+}
+
+async fn cmd_crucible_dataset_mark_non_provisionable(
```

Comment:
I know there's a `--destructive` flag for db ops, should that gate this operation?

(and the `mark_provisionable` variant)